Ls l

reckyy opened the pull request:
ご確認の程、よろしくお願いいたします！

reckyy committed:
ls -lオプション追加

既存のlsメソッドとの兼ね合いは、全く考えていなかった。
一旦、別途ls_lメソッドを用意して、作ってしまう。

reckyy committed:
parse_fileメソッドを変更

後々オプションを合体させるにあたって、parse_file内で完結させるべきだと
思ったので、変更した。

reckyy committed:
ファイルタイプの抜き出しに成功

reckyy committed:
パーミッション抜き出し成功

reckyy committed:
file_listの表示方法変更

reckyy committed:
ハードリンク抜き出し成功

reckyy committed:
以降、抜き出し成功したらその対象名のみ記す

reckyy committed:
オーナー名

reckyy committed:
グループ名

reckyy committed:
ファイルサイズ

reckyy committed:
最終アクセス日時

reckyy committed:
ファイル名

reckyy committed:
convert_to_file_permissionメソッドを修正

fs.modeが7桁の場合、抽出がずれてしまうので、末尾3桁を抽出するように修正

reckyy committed:
lsコマンドに見た目を合わせた

reckyy committed:
実際のlsに合わせて空白を調整

reckyy committed:
atimeではなく、mtimeに修正

reckyy committed:
表示ロジックをメソッドに分離

ついでに合計バイトサイズの表示も追加

reckyy committed:
rubocop修正

all_filesへの代入を条件分岐から外した

reckyy committed:
rubocop修正

reckyy committed:
rubocop修正

resize_file_listのメソッドが複雑すぎると怒られたので、
細分化した。

betachelsea commented on 04.ls/ls.rb:
```diff
@@ -27,6 +34,99 @@ def ls(all_files, total_row, width)
+    socket: 's',
+    unknown: 'u'
+  }
+  all_file_type[file_stat.ftype.to_sym]
```
[IMO] 書き方の好みが人によるかなという感じですが、こう書いてしまうと変数不要になりそうです 💡 

```ruby
def convert_to_file_type(file_stat)
  {
    file: '-',
    directory: 'd',
    characterSpecial: 'c',
    blockSpecial: 'b',
    fifo: 'p',
    link: 'l',
    socket: 's',
    unknown: 'u'
  }[file_stat.ftype.to_sym]
end
```

reckyy replied:
わざわざ変数に入れる意味がなかったですね。。
修正いたしました！

betachelsea commented on 04.ls/ls.rb:
```diff
@@ -27,6 +34,99 @@ def ls(all_files, total_row, width)
+    6 => 'rw-',
+    7 => 'rwx'
+  }
+  file_permission_type = ''
+  file_permission_number.each_char do |c|
+    file_permission_type += file_permission_string[c.to_i]
+  end
+  file_permission_type
```
[nits] これは、file_permission_number の 3文字を変換している様子ですね。 文字列処理は `+=` で繋げるよりも、次のようにするとRubyっぽいかなと思いました〜 👀 

```ruby
number = "755"
number.chars.map{|c| file_permission_string[c.to_i] }.join
# 文字列を配列に一旦入れて join でくっつける…というのはよくやりがちな気がします
# Rubyの文字列操作で += は あんまりやらずに済ませようとするケースが多いと思います。
# たとえばこんな感じとか…

puts "numberは" + number + "です" # こうしない
puts "numberは#{number}です" # こう書きがち
```

reckyy replied:
確かに変数も用意してるし、それを繋ぎ合わせて三度手間になってますね、、
`each_char`メソッドはEmuratorオブジェクトを返すため、mapが使えるということですね、、
修正しました！

betachelsea commented on 04.ls/ls.rb:
```diff
@@ -27,6 +34,99 @@ def ls(all_files, total_row, width)
+  total_blocks = get_file_stat(all_files, file_list)
+  file_list = resize_file_list(file_list)
+  print_file_list(file_list, total_blocks)
+end
```
[MUST] 引数で渡した配列を `file_list` に値を投入する、という形になっていますが、引数をメソッド内で変更してしまうと(確かにたまにそういう挙動をするメソッドもありますが)、ちょっと利用側としてはビックリしてしまいそうだと思いました。大抵の場合、引数として渡した値は変更無し、処理後の値は戻り値として得られる、という挙動が期待されるケースの方が多いかと思います。
今回の `get_file_stat` の呼び出しの様子をみると、 `file_list` としては空の配列だけ渡しているという感じなので、引数として貰わずに済ませても良さそうです。やるとしたら、`calculate_row_and_space` のように2つの戻り値を出力するようにしたほうが、コードを読んでいての驚き感は少ない気がします。 ちょっと再検討してみていただけますでしょうか 🙏 

```ruby
# コードを上から順に読んでいて思ったこと

def ls_l(all_files)
  file_list = [] # おっ、空の配列だ。
  total_blocks = get_file_stat(all_files, file_list) # ブロックの合計を計算しているんだな。どうして空の配列を渡すんだろう。
  file_list = resize_file_list(file_list) # どうして空の配列を渡すんだろう(再)、あとなぜか上書きしちゃってるな。 
  print_file_list(file_list, total_blocks) # 何かしら出力するための情報がfile_listに入ってそうだが、どのタイミングでfile_listは空の配列では無くなったのだろう。
end

```

reckyy replied:
何も考えず、行なってしまっていました。。
確かに、もう決まっている値を引数に今まで入れていたのに。。失敗です😭
修正いたしました！

betachelsea commented on 04.ls/ls.rb:
```diff
@@ -27,6 +34,99 @@ def ls(all_files, total_row, width)
+  before_resize_file_list = file_list.transpose
+  before_resize_file_list.each_with_index do |bfl, i|
+    max = calculate_max(bfl, i)
+    adjust_elements(bfl, max)
```
[IMO] この `calculate_max` と `adjust_elements` は、見た感じ不可分だし、うまくやったらまとめられそうだな〜と感じました :eyes:  どちらも `bfl.first.is_a?(String)` で判定してたり するあたりが特にそう感じるポイントかもしれません。

reckyy replied:
確かに、calculate_maxの過程はadjust_elementsメソッドの中でできますね。。これは共通点が明確で自分で修正しないといけない部分でした😭

betachelsea commented on 04.ls/ls.rb:
```diff
@@ -27,6 +34,99 @@ def ls(all_files, total_row, width)
+    file_stat << file_mtime.day
+    file_stat << file_mtime.strftime('%H:%M')
+    file_stat << file
+    file_list << file_stat
```
[nits] やや、やむを得ないところもあるかもしれませんが、 `file_stat` `file_list` `file_mtime` とほぼ字面が同じなので、この辺りを読むのがかなり難しいな〜と感じました。
最終的には、`file_list` に配列を追加したいということのようなので、変数を使わずに直接配列定義して入れちゃっても良いかなと思いました 💡 

```ruby
# すみません、動作確認していませんが、いけるかと…！
file_list = []
file_list << [
  "abc",
  "def",
  "ghi",
]

# =>  [["abc", "def", "ghi"]]
```

reckyy replied:
これも、変数名が多いと自分でも思っていました。。
修正いたしました！

betachelsea reviewed (commented):
コメントしました〜 📝

reckyy committed:
convert_to_file_typeメソッドを変更

一度変数に入れてから、変数[symbol]で呼び出していたが、
ただhash[symbol]としたいだけなので、変数を削除

reckyy committed:
convert_to_file_permissionメソッド変更

これも一度変数に入れていたのを直接戻り値に。
あと、文字列を繋げるのに += でしてしまっていた。これをmapで配列に入れ、joinで繋げるように変更

reckyy committed:
get_file_statメソッドを変更

file_listを各行ごとの2次元配列にするために、一度file_statという配列を用意して
それをfile_listに入れる、としていたが先のcommitと同じで直接入れることに。;

reckyy committed:
adjust_elementsとcalculate_maxをまとめた

reckyy committed:
file_listを引数で渡さないように変更

betachelsea commented on 04.ls/ls.rb:
```diff
@@ -27,6 +34,79 @@ def ls(all_files, total_row, width)
+
+def adjust_elements(bfl, idx)
+  if bfl.first.is_a?(String)
+    max = bfl.max_by(&:length).length
```
[IMO] ほぼ同じ意味合いの変数をちょっと加工したいケースで、次のように書かれている様子が多そうだなと感じました。

```
max = 一度なんらかの値を作る
max = maxを加工する操作
```

ここでどちらもほぼ「max」という値なので、maxとしか言いようがない気持ちもわかるのですが、同じ変数名に意味合いのやや異なる値が投入されていることにコードを読んでいくなかで若干の難しさを感じました。うまく変数名でその差異(なぜ編集したかったのか)を表現できると、より操作の必要性について将来の閲覧者に訴えることができそうです。

うーん、今回で言うなら、just ideaですが、次のような命名をしても良いかなと思いました〜．参考まで！🙏

- 前者のmax → max のまま もしくは max_length
- 後者のmax → 最大値というより見栄え上の幅を想定していそうなので、いっそ width とか？

reckyy replied:
こちらも確かにその通りですね。。
特に、widthはlsメソッドで使用しているのに、ここで使わないのはおかしかったですね、修正しました！

betachelsea commented on 04.ls/ls.rb:
```diff
@@ -27,6 +34,79 @@ def ls(all_files, total_row, width)
+def get_file_stat(all_files)
+  file_list = []
+  total_blocks = 0
+  all_files.each do |file|
```
[nits] 空配列の定義と、そこに代入するeachの操作があると、まとめられそうだな〜と感じます！

```ruby
file_list = all_files.map do …
```

すみません、動作確認はできてませんが、いけそうな気がしたので共有です。ダメだったらスルーでOKです！🙏

reckyy replied:
いけました☺️ありがとうございます！

betachelsea commented on 04.ls/ls.rb:
```diff
@@ -27,6 +34,79 @@ def ls(all_files, total_row, width)
+
+def convert_to_file_permission(file_stat)
+  file_permission_number = file_stat.mode.to_s(8).slice(-3..-1)
+  file_permission_string = {
```
[nits] 改めて見ると、決まった値なので、定数として扱ってしまっても良さそうですね👀

reckyy replied:
こちらだけ、先に返信させていただきます！
定数として扱うというのは、以下のイメージで正しいでしょうか。
```ruby
#定数はトップレベル（もしくはメソッド内）で定義して、convert_to_file_typeメソッドを削除
FILE_TYPE = {
    file: '-',
    directory: 'd',
    characterSpecial: 'c',
    blockSpecial: 'b',
    fifo: 'p',
    link: 'l',
    socket: 's',
    unknown: 'u'
  }
def get_file_stat(fs)
...略...
  file_list << [
...略...
    FILE_TYPE[fs.ftype.to_sym] + convert_to_file_permission(fs),
```

reckyy replied:
こちらもし認識が誤ってたらいけないので、据え置きでマージします！

betachelsea commented on 04.ls/ls.rb:
```diff
@@ -1,12 +1,19 @@
+    ls_l(all_files)
+  else
+    total_row, width = calculate_row_and_space(all_files)
+    ls(all_files, total_row, width)
```
[IMO] そういえば前回、懸念されていたメソッドの共存or分割？についてコメント失念しておりました！🙏 

ls と ls_l に分かれるぶんについては、かなり処理の内容も違うので、致し方ないかなと思います。強いて言えば、ふたつのメソッドが発生したことで、2つの観点から、メソッド名の見直しはあっても良いかもと思いました。
- このプログラムは全体で ls コマンドの動きを形作るので、`ls` という名前を内部のいちメソッドが持っているにふさわしいのかどうか。いま、`ls_l` と機能がほぼ二分されたことによって、そのメソッド単体を `ls` と呼んで良いかというのは再考の余地があるのではないか。 (考えた上でやっぱり主要機能を持たせているのだからok、としても良いとは思います) 
- `ls_l` という名前から、前提としてこのプログラムはlsおよびlongオプションを知っている人が見れば挙動は推測できるが、何にも知らないレビュアーが見た時はどう解釈するか。( エルって何の意味だろう？)

命名、難しすぎて自分も全然自信ないのですが😅 「無知識のレビュアーがパッと見た時に意味に気付くことができるか？」は意識してやると、未来の自分(N年後にメンテナンスするように指示されたけど昔のことはキレイサッパリ忘れている)を救うことにもつながると思います！プラクティスのなかで今後もコードの見直しは無限回やると思いますが、その時に「なんにも知らない・覚えていない自分」をインストール(いや、初期化というべきでしょうか)して確認するのをぜひやってみていただきたいです〜👍

reckyy replied:
いえいえ、とんでもないです！
返信ありがとうございます 🙇 
結論からお話しすると、後々lsメソッドのオプションを複数追加することから、
`ls`メソッド -> `ls_no_opt`メソッド
`ls_l`メソッド-> `ls_opt_l`メソッド
に変更しました🧐

> 命名、難しすぎて自分も全然自信ないのですが😅 「無知識のレビュアーがパッと見た時に意味に気付くことができるか？」は意識してやると、未来の自分(N年後にメンテナンスするように指示されたけど昔のことはキレイサッパリ忘れている)を救うことにもつながると思います！プラクティスのなかで今後もコードの見直しは無限回やると思いますが、その時に「なんにも知らない・覚えていない自分」をインストール(いや、初期化というべきでしょうか)して確認するのをぜひやってみていただきたいです〜👍

そうですよね、、プラクティスのレビューが帰ってきた時点でさえメソッドの働きをすでに忘れている自分がいますからね、、自分で作ったのに😭
より意識していきたいと思います！

betachelsea reviewed (commented):
修正ありがとうございます！確認OKです👍少し補足コメントしました！

reckyy committed:
adjust_elementsの変数名を変更

reckyy committed:
file_listに直接値を代入

reckyy committed:
ls,ls_lをより直感的なメソッド名に変更
